funcx_sdk/funcx/sdk/asynchronous/ws_polling_task.py

```diff
@@ -158,70 +159,87 @@ async def handle_incoming(self, pending_futures, auto_close=False) -> bool:
+                    return True
 
             # Handle the results received but not processed before
             unprocessed_task_ids = self.unknown_results.keys() & pending_futures.keys()
             for task_id in unprocessed_task_ids:
+                task_fut = pending_futures.pop(task_id)
                 data = self.unknown_results.pop(task_id)
-                if await self.set_result(task_id, data, pending_futures):
+                if await self.set_result(task_fut, data):
                     return True
 
         log.info("WebSocket connection closed by main thread")
         return True
 
-    async def set_result(self, task_id, data, pending_futures):
+    async def set_result(self, task_fut: FuncXFuture, task_data: t.Dict):
         """Sets the result of a future with given task_id in the pending_futures map,
         then decrement the atomic counter and close the WebSocket connection if needed
 
         Parameters
         ----------
-        task_id : str
-            Task ID of the future to set the result for
+        task_fut : FuncXFuture
+            Task future for which to parse and set task_data
 
-        data : dict
-            Dict containing result/exception that should be set to future
 
-        pending_futures : dict
-            Dict of task_id keys that map to their futures
+        task_data : dict
+            Dict containing result/exception that should be set to future
 
         Returns
         -------
         bool
             True if the WebSocket connection has closed and the thread can
             exit, False otherwise
         """
-        future = pending_futures.pop(task_id)
         try:
-            if data["result"]:
-                future.set_result(
-                    self.funcx_client.fx_serializer.deserialize(data["result"])
+            if "result" in task_data:
+                task_fut.set_result(
+                    self.funcx_client.fx_serializer.deserialize(task_data["result"])
                 )
-            elif data["exception"]:
+            elif "exception" in task_data:
                 r_exception = self.funcx_client.fx_serializer.deserialize(
-                    data["exception"]
+                    task_data["exception"]
                 )
-                future.set_exception(dill.loads(r_exception.e_value))
+                task_fut.set_exception(dill.loads(r_exception.e_value))
             else:
-                future.set_exception(Exception(data["reason"]))
-        except Exception:
-            log.exception("Caught unexpected exception while setting results")
-
-        # When the counter hits 0 we always exit. This guarantees that that if the
-        # counter increments to 1 on the executor, this handler needs to be restarted.
+                st, reason = task_data.get("status"), task_data.get("reason")
+                msg = f"Status: {st}.  Reason: {reason}"
+                task_fut.set_exception(Exception(msg))
```

Comment:
I'd like to look at these side-by-side for comparison. I think the fewer variables we need to track, the better.

```suggestion
                st, reason = task_data.get("status"), task_data.get("reason")
                task_fut.set_exception(Exception(f"Status: '{st}', Reason: '{reason}'"))
```

I even like

```suggestion
                task_fut.set_exception(
                    Exception(f"Data contained neither result nor exception: {task_data}")
                )
```

Reply:
No, this makes sense.  The latter is a better call -- this section has changed a smidgen since I first added this in #727.  I'll add that directly.